include/multipass/exceptions/base_qexception.h

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (C) Canonical, Ltd.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 3.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#ifndef MULTIPASS_BASE_QEXCEPTION
+#define MULTIPASS_BASE_QEXCEPTION
+
+#include <QException>
+
+#include <string>
+
+namespace multipass
+{
+// CRTP template base class that is the boilerplate code of QException derived classes
+template <typename DerivedException>
+class BaseQException : public QException
+{
+public:
+    BaseQException(const std::string& err) : error_string{err}
+    {
+        // use concepts instead of static_assert or enable_if to apply the contraint once C++ 20 arrives.
+        static_assert(std::is_base_of_v<BaseQException, DerivedException>,
+                      "DerivedException must be derived from BaseQException");
+    }
+
+    void raise() const override
+    {
+        throw static_cast<const DerivedException&>(*this);
+    }
+
+    BaseQException* clone() const override
+    {
+        return new DerivedException(static_cast<const DerivedException&>(*this));
```

Comment:
Hey @georgeliao, while this is a very clever idea, it is a little dangerous. It works well for the first derived class, but it will [break on the second](https://godbolt.org/z/b9E5PfvjE).

That said, I am fine with keeping it as long as:
- we document that this template should be instantiated with the ultimate concrete type for `DerivedException` (not an intermediate type)
- explain that means that any derived class should either be final, be itself templated, or override `raise` and `clone` (which defeats the purpose)
- follow that recipe in any derived exception today

Reply:
That is nice the observation of CRTP, I think generally CRTP is used in one level of inheritance. So labeling the derived classes to final and document it should be more appropriate.